From 97da81ff0e85d27cb928492c0ac84987e9f16de5 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Tue, 19 Jan 2016 15:46:00 +0800
Subject: [PATCH 0196/2241] clk: a3700: add clock gating driver for both south
 & north bridges

- Add ARMADA_3700_CLK configuration which also enable
  the common module of MVEBU_CLK_COMMON
- Update the clk_gating_soc_desc structure to support
  clk_gate_flags

Change-Id: I770925fe9cf1c812825ee5c1316a1139026bb59d
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/27816
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
---
 .../bindings/clock/mvebu-armada3700-clock.txt      | 40 ++++++++++++++++
 arch/arm64/Kconfig.platforms                       |  1 +
 arch/arm64/boot/dts/marvell/armada-37xx.dtsi       | 12 +++++
 drivers/clk/mvebu/Kconfig                          |  4 ++
 drivers/clk/mvebu/Makefile                         |  1 +
 drivers/clk/mvebu/armada-3700.c                    | 56 ++++++++++++++++++++++
 drivers/clk/mvebu/common.c                         |  2 +-
 drivers/clk/mvebu/common.h                         |  1 +
 8 files changed, 116 insertions(+), 1 deletion(-)
 create mode 100644 Documentation/devicetree/bindings/clock/mvebu-armada3700-clock.txt
 create mode 100644 drivers/clk/mvebu/armada-3700.c

diff --git a/Documentation/devicetree/bindings/clock/mvebu-armada3700-clock.txt b/Documentation/devicetree/bindings/clock/mvebu-armada3700-clock.txt
new file mode 100644
index 0000000..3c76d93
--- /dev/null
+++ b/Documentation/devicetree/bindings/clock/mvebu-armada3700-clock.txt
@@ -0,0 +1,40 @@
+* Clock bindings for Marvell MVEBU Armada3700 SoCs
+
+Marvell Armada 3700 allow some peripheral clocks to be gated to save power.The clock
+consumer should specify the desired clock by having the clock ID in its
+"clocks" phandle cell. The clock ID is directly mapped to the corresponding clock
+gating control bit in HW to ease manual clock lookup in datasheet.
+
+Required properties:
+- compatible : shall be one of the following:
+	"marvell,armada-3700-north-bridge-gating-clock" - gate clock control for Armada 3700 north bridge
+	"marvell,armada-3700-south-bridge-gating-clock" - gate clock control for Armada 3700 south bridge
+- reg : shall be the register address of the Clock Gating Control register
+- #clock-cells : from common clock binding; shall be set to 1
+
+Optional properties:
+- clocks : default parent clock phandle and clock specifier.
+	(for gate clock control, it's optional to define the parent clock,
+	for the clock consumer, it defines the gate clock phandle and clock ID)
+
+Example:
+
+ngateclk: north-bridge-gate-clk@13014 {
+	compatible = "marvell,armada-3700-north-bridge-gating-clock";
+	reg = <0x13014 0x4>;
+	#clock-cells = <1>;
+};
+
+sgateclk: south-bridge-gate-clk@18014 {
+	compatible = "marvell,armada-3700-south-bridge-gating-clock";
+	reg = <0x18014 0x4>;
+	#clock-cells = <1>;
+};
+
+sata0: sata@e0000 {
+	compatible = "marvell,armada-3700-ahci";
+	reg = <0xe0000 0x2000>;
+	interrupts = <0 27 4>;
+	clocks = <&ngateclk 3>;
+	status = "disabled";
+};
\ No newline at end of file
diff --git a/arch/arm64/Kconfig.platforms b/arch/arm64/Kconfig.platforms
index 575614d..6534337 100644
--- a/arch/arm64/Kconfig.platforms
+++ b/arch/arm64/Kconfig.platforms
@@ -5,6 +5,7 @@ config ARCH_MVEBU
 	select ARM_GIC_ODMI
 	select MVEBU_PIC
 	select MVEBU_CLK_COMMON
+	select ARMADA_3700_CLK
 	select ARMADA_AP806_CORE_CLK
 	select ARMADA_AP806_RING_CLK
 	select ARMADA_CP110_CLK
diff --git a/arch/arm64/boot/dts/marvell/armada-37xx.dtsi b/arch/arm64/boot/dts/marvell/armada-37xx.dtsi
index 91f7e01..fcbe697 100644
--- a/arch/arm64/boot/dts/marvell/armada-37xx.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-37xx.dtsi
@@ -118,6 +118,18 @@
 				status = "disabled";
 			};
 
+			ngateclk: north-bridge-gate-clk@13014 {
+				compatible = "marvell,armada-3700-north-bridge-gating-clock";
+				reg = <0x13014 0x4>;
+				#clock-cells = <1>;
+			};
+
+			sgateclk: south-bridge-gate-clk@18014 {
+				compatible = "marvell,armada-3700-south-bridge-gating-clock";
+				reg = <0x18014 0x4>;
+				#clock-cells = <1>;
+			};
+
 			usb3@58000 {
 				compatible = "generic-xhci";
 				reg = <0x58000 0x4000>;
diff --git a/drivers/clk/mvebu/Kconfig b/drivers/clk/mvebu/Kconfig
index 8d6f007..3206856 100644
--- a/drivers/clk/mvebu/Kconfig
+++ b/drivers/clk/mvebu/Kconfig
@@ -31,6 +31,10 @@ config ARMADA_XP_CLK
 	select MVEBU_CLK_CPU
 	select MVEBU_CLK_COREDIV
 
+config ARMADA_3700_CLK
+	bool
+	select MVEBU_CLK_COMMON
+
 config DOVE_CLK
 	bool
 	select MVEBU_CLK_COMMON
diff --git a/drivers/clk/mvebu/Makefile b/drivers/clk/mvebu/Makefile
index 8390f06..9f0d0b4 100644
--- a/drivers/clk/mvebu/Makefile
+++ b/drivers/clk/mvebu/Makefile
@@ -5,6 +5,7 @@ obj-$(CONFIG_ARMADA_370_CLK)	+= armada-370.o
 obj-$(CONFIG_ARMADA_375_CLK)	+= armada-375.o
 obj-$(CONFIG_ARMADA_38X_CLK)	+= armada-38x.o
 obj-$(CONFIG_ARMADA_39X_CLK)	+= armada-39x.o
+obj-$(CONFIG_ARMADA_3700_CLK)	+= armada-3700.o
 obj-$(CONFIG_ARMADA_AP806_CORE_CLK) += ap806-core.o
 obj-$(CONFIG_ARMADA_AP806_RING_CLK) += ap806-ring.o
 obj-$(CONFIG_ARMADA_CP110_CLK)  += cp110.o
diff --git a/drivers/clk/mvebu/armada-3700.c b/drivers/clk/mvebu/armada-3700.c
new file mode 100644
index 0000000..ada70a2
--- /dev/null
+++ b/drivers/clk/mvebu/armada-3700.c
@@ -0,0 +1,56 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#include <linux/kernel.h>
+#include <linux/clk-provider.h>
+#include <linux/io.h>
+#include <linux/of.h>
+#include "common.h"
+
+/*
+ * Clock Gating Control
+ */
+static const struct clk_gating_soc_desc armada_3700_north_bridge_gating_desc[] __initconst = {
+	{ "sata-host", NULL, 3, 0, CLK_GATE_SET_TO_DISABLE },
+	{ }
+};
+
+static void __init
+armada_3700_north_bridge_clk_gating_init(struct device_node *np)
+{
+	mvebu_clk_gating_setup(np, armada_3700_north_bridge_gating_desc);
+}
+CLK_OF_DECLARE(armada_3700_north_bridge_clk_gating,
+	       "marvell,armada-3700-north-bridge-gating-clock",
+	       armada_3700_north_bridge_clk_gating_init);
+
+static const struct clk_gating_soc_desc armada_3700_south_bridge_gating_desc[] __initconst = {
+	{ "gbe1-core", NULL, 4, 0, CLK_GATE_SET_TO_DISABLE },
+	{ "gbe0-core", NULL, 5, 0, CLK_GATE_SET_TO_DISABLE },
+	{ "usb32-ss-sys", NULL, 17, 0, CLK_GATE_SET_TO_DISABLE },
+	{ }
+};
+
+static void __init
+armada_3700_south_bridge_clk_gating_init(struct device_node *np)
+{
+	mvebu_clk_gating_setup(np, armada_3700_south_bridge_gating_desc);
+}
+CLK_OF_DECLARE(armada_3700_south_bridge_clk_gating,
+	       "marvell,armada-3700-south-bridge-gating-clock",
+	       armada_3700_south_bridge_clk_gating_init);
diff --git a/drivers/clk/mvebu/common.c b/drivers/clk/mvebu/common.c
index 4a353cb..45e1b98c 100644
--- a/drivers/clk/mvebu/common.c
+++ b/drivers/clk/mvebu/common.c
@@ -294,7 +294,7 @@ void __init mvebu_clk_gating_setup(struct device_node *np,
 			(desc[n].parent) ? desc[n].parent : default_parent;
 		ctrlp->gates[n] = clk_register_gate(NULL, desc[n].name, parent,
 					desc[n].flags, base, desc[n].bit_idx,
-					0, ctrlp->lock);
+					desc[n].clk_gate_flags, ctrlp->lock);
 		WARN_ON(IS_ERR(ctrlp->gates[n]));
 	}
 
diff --git a/drivers/clk/mvebu/common.h b/drivers/clk/mvebu/common.h
index f0de6c8..d820b5d 100644
--- a/drivers/clk/mvebu/common.h
+++ b/drivers/clk/mvebu/common.h
@@ -42,6 +42,7 @@ struct clk_gating_soc_desc {
 	const char *parent;
 	int bit_idx;
 	unsigned long flags;
+	u8 clk_gate_flags;
 };
 
 void __init mvebu_coreclk_setup(struct device_node *np,
-- 
2.7.4

